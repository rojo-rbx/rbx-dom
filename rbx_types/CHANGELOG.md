# rbx_types Changelog

## Unreleased Changes
* Add `len` and `is_empty` methods to `Attributes` struct. ([#377])

## 1.7.0 (2023-10-03)
* Implemented `FromStr` for `TerrainMaterials`. ([#354])
* `MaterialColorsError` and `UniqueIdError` are no longer publicly exposed. ([#355])
* Implemented barebones `SecurityCapabilities`. ([#358])
* Implement `Display` for `SharedStringHash` ([#363])

[#354]: https://github.com/rojo-rbx/rbx-dom/pull/354
[#355]: https://github.com/rojo-rbx/rbx-dom/pull/355
[#358]: https://github.com/rojo-rbx/rbx-dom/pull/358
[#363]: https://github.com/rojo-rbx/rbx-dom/pull/363
[#377]: https://github.com/rojo-rbx/rbx-dom/pull/377

## 1.6.0 (2023-08-09)
* Added support for `UniqueId` values. ([#271])
* Changed `BinaryString`'s non-human readable serde implementation to be identical to `Vec<u8>`. ([#276])
* Added `Font::new` and `Font::regular` constructors. ([#283])
* Added support for `CFrame` values in attributes. ([#296])
* Added support for `Font` values in attributes. ([#299])
* Implemented `MaterialColors`. ([#323])

[#271]: https://github.com/rojo-rbx/rbx-dom/pull/271
[#276]: https://github.com/rojo-rbx/rbx-dom/pull/276
[#283]: https://github.com/rojo-rbx/rbx-dom/pull/283
[#296]: https://github.com/rojo-rbx/rbx-dom/pull/296
[#299]: https://github.com/rojo-rbx/rbx-dom/pull/299
[#323]: https://github.com/rojo-rbx/rbx-dom/pull/323

## 1.5.0 (2023-04-22)
* Implemented `Font`. ([#248])

[#248]: https://github.com/rojo-rbx/rbx-dom/pull/248

## 1.4.2 (2022-06-12)
* `Variant::String` now encodes correctly inside of `Attributes`.

## 1.4.1 (2022-06-12)
* `Attributes` can now decode from an empty buffer.
* `Attributes` now encodes as an empty buffer when empty.

## 1.4.0 (2022-06-05)
* Added `Attributes::with` for creating attributes in code more easily.
* Implemented `Hash` for `VariantTy`.

## 1.3.0 (2021-10-11)
* Implemented `Tags`. ([#199])
* Implemented `Attributes`. ([#166])

[#166]: https://github.com/rojo-rbx/rbx-dom/pull/166
[#199]: https://github.com/rojo-rbx/rbx-dom/pull/199

## 1.2.0 (2021-07-19)
* Implemented `From<Color3>` for `Color3uint8` and `From<Color3uint8>` for `Color3`. ([#198][#198])

[#198]: https://github.com/rojo-rbx/rbx-dom/pull/198

## 1.1.0 (2021-07-02)
* Critical fix: changed serde serialization of fields from PascalCase to camelCase. ([#191][#191])

[#191]: https://github.com/rojo-rbx/rbx-dom/pull/191

## 1.0.0 (2021-06-26) (yanked)
* Removed `legacy-compact` feature, which added conversions from rbx\_dom\_weak 1.x types.

## 0.4.0 (2021-06-26)
* Changed `Variant` serialization to use Serde's default enum representation.
  * Before: `{ "type": "Vector2", "value": [1.0, 2.0] }`
  * After: `{ "Vector2": [1.0, 2.0] }`
* Implemented `PartialOrd` and `Ord` for `SharedStringHash`.

## 0.3.1 (2021-05-14)
* Added `Variant::OptionalCoordinateFrame`.

## 0.3.0 (2021-02-16)
* Renamed `EnumValue` to `Enum`.
* Added `Display` and `FromStr` implementations for `Ref`.

## 0.2.0 (2020-04-27)
* `Ref` can now represent null explicitly via `Ref::none` and `Ref::is_none`.
* Added `Region3` and `Region3int16` to `Variant` and `VariantType`.
* Added `legacy-compat` feature to enable conversions with rbx_dom_weak 1.x.

## 0.1.0 (2020-02-05)
* Initial release of types crate, should have rough feature parity with rbx_dom_weak.
* API will move a lot before becoming stable.
