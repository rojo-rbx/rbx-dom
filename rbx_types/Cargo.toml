[package]
name = "rbx_types"
description = "Types used to represent Roblox values"
version = "1.4.2"
authors = ["Lucien Greathouse <me@lpghatguy.com>"]
edition = "2018"
license = "MIT"
readme = "README.md"
homepage = "https://github.com/rojo-rbx/rbx-dom"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
base64 = "0.13.0"
blake3 = "1.3.1"
bitflags = "1.3.2"
lazy_static = "1.4.0"
rand = "0.8.5"
thiserror = "1.0.31"
serde = { version = "1.0.137", features = ["derive"], optional = true }

[dev-dependencies]
insta = { version="1.14.1", features=["yaml"] }
bincode = "1.3.3"
serde_json = "1.0.81"
