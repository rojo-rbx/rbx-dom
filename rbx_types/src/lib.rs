#[cfg(feature = "serde")]
#[macro_use]
mod serde_util;

mod axes;
mod basic_types;
mod binary_string;
mod brick_color;
mod content;
mod faces;
mod lister;
mod physical_properties;
mod referent;
mod shared_string;
mod tags;
mod variant;

pub use axes::*;
pub use basic_types::*;
pub use binary_string::*;
pub use brick_color::*;
pub use content::*;
pub use faces::*;
pub use physical_properties::*;
pub use referent::*;
pub use shared_string::*;
pub use tags::*;
pub use variant::*;
