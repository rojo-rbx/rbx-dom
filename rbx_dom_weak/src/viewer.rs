use std::{
    collections::{BTreeMap, HashMap},
    fmt::Write,
};

use crate::{
    types::{Ref, Variant},
    WeakDom,
};
use rbx_types::RobloxString;
use serde::{Deserialize, Serialize};
use ustr::Ustr;

/// Contains state for viewing and redacting nondeterministic portions of
/// WeakDom objects, making them suitable for usage in snapshot tests.
///
/// `DomViewer` can be held onto and used with a DOM multiple times. IDs will
/// persist when viewing the same instance multiple times, and should stay the
/// same across multiple runs of a test.
pub struct DomViewer {
    referent_to_id: HashMap<Ref, String>,
    next_id: usize,
}

impl DomViewer {
    /// Construct a new `DomViewer` with no interned referents.
    pub fn new() -> Self {
        Self {
            referent_to_id: HashMap::new(),
            next_id: 0,
        }
    }

    /// View the given `WeakDom`, creating a `ViewedInstance` object that can be
    /// used in a snapshot test.
    pub fn view(&mut self, dom: &WeakDom) -> ViewedInstance {
        let root_referent = dom.root_ref();
        self.populate_referent_map(dom, root_referent);
        self.view_instance(dom, root_referent)
    }

    /// View the children of the root instance of the given `WeakDom`, returning
    /// them as a `Vec<ViewedInstance>`.
    pub fn view_children(&mut self, dom: &WeakDom) -> Vec<ViewedInstance> {
        let root_instance = dom.root();
        let children = root_instance.children();

        for &referent in children {
            self.populate_referent_map(dom, referent);
        }

        children
            .iter()
            .map(|&referent| self.view_instance(dom, referent))
            .collect()
    }

    fn populate_referent_map(&mut self, dom: &WeakDom, referent: Ref) {
        let next_id = &mut self.next_id;
        self.referent_to_id.entry(referent).or_insert_with(|| {
            let name = format!("referent-{}", next_id);
            *next_id += 1;
            name
        });

        let instance = dom.get_by_ref(referent).unwrap();
        for referent in instance.children() {
            self.populate_referent_map(dom, *referent);
        }
    }

    fn view_instance(&self, dom: &WeakDom, referent: Ref) -> ViewedInstance {
        let instance = dom.get_by_ref(referent).unwrap();

        let children = instance
            .children()
            .iter()
            .copied()
            .map(|referent| self.view_instance(dom, referent))
            .collect();

        let properties = instance
            .properties
            .iter()
            .map(|(key, value)| {
                let new_value = match value {
                    Variant::Ref(referent) => {
                        if referent.is_some() {
                            let referent_str = self
                                .referent_to_id
                                .get(referent)
                                .cloned()
                                .unwrap_or_else(|| "[unknown ID]".to_owned());

                            ViewedValue::Ref(referent_str)
                        } else {
                            ViewedValue::Ref("null".to_owned())
                        }
                    }
                    Variant::SharedString(shared_string) => {
                        let hash = shared_string.hash();
                        let mut hash_hex = String::with_capacity(hash.as_bytes().len() * 2);

                        for byte in hash.as_bytes() {
                            write!(hash_hex, "{:02x}", byte).unwrap();
                        }
                        ViewedValue::SharedString {
                            len: shared_string.data().len(),
                            hash: hash_hex,
                        }
                    }
                    other => ViewedValue::Other(other.clone()),
                };

                (*key, new_value)
            })
            .collect();

        ViewedInstance {
            referent: self.referent_to_id.get(&referent).unwrap().clone(),
            name: instance.name.clone(),
            class: instance.class,
            properties,
            children,
        }
    }
}

impl Default for DomViewer {
    fn default() -> Self {
        Self::new()
    }
}

/// A transformed view into a `WeakDom` or `Instance` that has been redacted and
/// transformed to be more readable.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ViewedInstance {
    referent: String,
    name: RobloxString,
    class: Ustr,
    properties: BTreeMap<Ustr, ViewedValue>,
    children: Vec<ViewedInstance>,
}

/// Wrapper around Variant with refs replaced to be redacted, stable versions of
/// their original IDs.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
enum ViewedValue {
    Ref(String),
    SharedString { len: usize, hash: String },
    Other(Variant),
}

#[cfg(test)]
mod test {
    use super::*;

    use crate::types::SharedString;
    use crate::InstanceBuilder;

    #[test]
    fn redact_single() {
        let dom = WeakDom::new(InstanceBuilder::new("Folder").with_name("Root"));

        insta::assert_yaml_snapshot!(DomViewer::new().view(&dom));
    }

    #[test]
    fn redact_multi() {
        let dom = WeakDom::new(
            InstanceBuilder::new("Folder")
                .with_name("Root")
                .with_children(
                    (0..4)
                        .map(|i| InstanceBuilder::new("Folder").with_name(format!("Child {}", i))),
                ),
        );

        insta::assert_yaml_snapshot!(DomViewer::new().view(&dom));
    }

    #[test]
    fn redact_values() {
        let root = InstanceBuilder::new("ObjectValue").with_name("Root");
        let root_ref = root.referent;
        let root = root.with_property("Value", root_ref);

        let dom = WeakDom::new(root);

        insta::assert_yaml_snapshot!(DomViewer::new().view(&dom));
    }

    #[test]
    fn abbreviate_shared_string() {
        let shared_string = SharedString::new("foo".into());

        let root = InstanceBuilder::new("UnionOperation")
            .with_name("Root")
            .with_property("PhysicalConfigData", shared_string);

        let dom = WeakDom::new(root);

        insta::assert_yaml_snapshot!(DomViewer::new().view(&dom));
    }
}
