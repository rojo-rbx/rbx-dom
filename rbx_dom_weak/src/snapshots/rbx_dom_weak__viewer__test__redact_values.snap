---
source: rbx_dom_weak/src/viewer.rs
expression: "DomViewer::new().view(&dom)"
---
referent: referent-0
name:
  Utf8:
    - Root
class: ObjectValue
properties:
  Value: referent-0
children: []
